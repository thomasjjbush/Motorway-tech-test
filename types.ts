import { FormEvent, ReactNode } from 'react';
import HtmlWebPackPlugin from 'html-webpack-plugin';
import { Json } from 'enzyme-to-json';

export interface AspectRatioProps {
    children: ReactNode;
    ratio: AspectRatios;
}

export enum AspectRatios {
    '1:1' = 100,
    '16:9' = 56.25,
    '4:3' = 75,
}

export interface ButtonProps {
    children: string;
    disabled?: boolean;
    onClick?: () => void;
    type?: 'button' | 'submit' | 'reset';
}

export interface FormField {
    label: string;
    max?: number;
    min?: number;
    placeholder?: string;
    type?: string;
    valid?: boolean;
    validation?: RegExp;
    value?: number | string;
}

export interface FormProps {
    ariaLabel: string;
    fields: {
        [field: string]: FormField;
    };
    onSubmit: (e: FormEvent) => void;
}

export interface Image {
    id: string;
    created_at: string;
    updated_at: string;
    color: string;
    description: string;
    alt_description: string;
    categories: string[];
    likes: number;
    user: {
        id: string;
        updated_at: string;
        username: string;
        name: string;
        first_name: string;
        last_name: string;
        bio: string;
        location: string;
        profile_image: string;
        total_collections: number;
        total_likes: number;
        total_photos: number;
    };
    url: string;
}

export interface ImageProps {
    alt: string;
    round?: boolean;
    src: string;
    width?: string;
}

export interface ImagesProps {
    columns: number;
}

export interface SerializerMap extends Json {
    node: {
        type: {
            componentStyle?: {
                baseStyle?: {
                    rules: (string | ((props: StyledProps) => string))[];
                };
                rules: (string | ((props: StyledProps) => string))[];
            };
        };
    };
}

export type StyledProps<P = unknown> = { theme: Theme } & P;

export interface Theme {
    colors: {
        blue: string;
        red: string;
        yellow: string;
    };
}

export interface WebpackConfig {
    devServer: { historyApiFallback: boolean; open: boolean };
    entry: string[];
    mode: string;
    module: {
        rules: WebpackRule[];
    };
    output: { path?: string; publicPath: string };
    plugins: any[];
    resolve: { extensions: string[] };
}

interface WebpackRule {
    exclude?: (RegExp | string)[];
    test: RegExp;
    use: {
        loader: string;
        options?: { plugins: string[] };
    };
}
