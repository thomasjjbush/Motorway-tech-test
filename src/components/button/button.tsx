import React, { FC, ReactElement } from 'react';
import styled from 'styled-components';
import { ButtonProps as Props, StyledProps } from '../../../types';

const StyledButton = styled.button<StyledProps>`
    appearance: none;
    background-color: ${({ theme }) => theme.colors.blue};
    border: 0;
    border-radius: 5px;
    color: white;
    cursor: pointer;
    font-size: 16px;
    padding: 10px 20px;

    &[disabled] {
        cursor: not-allowed;
        opacity: 0.5;
    }
`;

export const Button: FC<Props> = ({ children, disabled = false, onClick, type = 'button' }: Props): ReactElement => (
    <StyledButton disabled={disabled} onClick={onClick} type={type}>
        {children}
    </StyledButton>
);
