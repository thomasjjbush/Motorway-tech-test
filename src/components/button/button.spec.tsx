import React from 'react';
import { shallow } from 'enzyme';
import { Button } from './button';
import { ButtonProps as Props } from '../../../types';

describe('button', () => {
    let props: Props;

    beforeEach(() => {
        props = {
            children: 'hello!',
            onClick: jest.fn(),
        };
    });

    it('should match snapshot', () => {
        expect(shallow(<Button {...props} />)).toMatchSnapshot();
    });

    it('should pass specified type prop correctly', () => {
        expect(shallow(<Button {...props} type="submit" />).prop('type')).toBe('submit');
    });

    it('should pass specified disabled prop correctly', () => {
        expect(shallow(<Button {...props} disabled />).prop('disabled')).toBe(true);
    });

    it('should invoke onClick cb on click', () => {
        shallow(<Button {...props} />).simulate('click');
        expect(props.onClick).toHaveBeenCalledTimes(1);
    });
});
