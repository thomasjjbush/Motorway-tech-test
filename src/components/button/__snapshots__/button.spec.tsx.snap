// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button should match snapshot 1`] = `
<styled.button
  disabled={false}
  onClick={[MockFunction]}
  styles="
    appearance: none;
    background-color: theme.colors.blue;
    border: 0;
    border-radius: 5px;
    color: white;
    cursor: pointer;
    font-size: 16px;
    padding: 10px 20px;

    &[disabled] {
        cursor: not-allowed;
        opacity: 0.5;
    }
"
  type="button"
>
  hello!
</styled.button>
`;
